crossbeam-queue/src/spsc.rs

```diff
@@ -0,0 +1,449 @@
+        }
+    }
+
+    /// Returns the distance between two positions.
+    ///
+    /// Positions must be in range `0 .. 2 * cap`.
+    #[inline]
+    fn distance(&self, a: usize, b: usize) -> usize {
+        if a <= b {
+            b - a
+        } else {
+            2 * self.cap - a + b
+        }
+    }
+}
+
+impl<T> Drop for Inner<T> {
+    fn drop(&mut self) {
+        let mut head = self.head.load(Ordering::Relaxed);
+        let tail = self.tail.load(Ordering::Relaxed);
+
+        // Loop over all slots that hold a value and drop them.
+        while head != tail {
+            unsafe {
+                self.slot(head).drop_in_place();
+            }
+            head = self.increment(head);
+        }
+
+        // Finally, deallocate the buffer, but don't run any destructors.
+        unsafe {
+            Vec::from_raw_parts(self.buffer, 0, self.cap);
+        }
+    }
+}
+
+/// Creates a bounded single-producer single-consumer queue with the given capacity.
+///
+/// Returns the producer and the consumer side for the queue.
+///
+/// # Panics
+///
+/// Panics if the capacity is zero.
+///
+/// # Examples
+///
+/// ```
+/// use crossbeam_queue::spsc;
+///
+/// let (p, c) = spsc::<i32>(100);
+/// ```
+pub fn spsc<T>(cap: usize) -> (Producer<T>, Consumer<T>) {
+    assert!(cap > 0, "capacity must be non-zero");
+
+    // Allocate a buffer of length `cap`.
+    let buffer = {
+        let mut v = Vec::<T>::with_capacity(cap);
+        let ptr = v.as_mut_ptr();
+        mem::forget(v);
+        ptr
+    };
+
+    let inner = Arc::new(Inner {
+        head: CachePadded::new(AtomicUsize::new(0)),
+        tail: CachePadded::new(AtomicUsize::new(0)),
+        buffer,
+        cap,
+        _marker: PhantomData,
+    });
+
+    let p = Producer {
+        inner: inner.clone(),
+        head: Cell::new(0),
+        tail: Cell::new(0),
+    };
+
+    let c = Consumer {
+        inner,
+        head: Cell::new(0),
+        tail: Cell::new(0),
+    };
+
+    (p, c)
+}
+
+/// The producer side of a bounded single-producer single-consumer queue.
+///
+/// # Examples
+///
+/// ```
+/// use crossbeam_queue::{spsc, PushError};
+///
+/// let (p, c) = spsc::<i32>(1);
+///
+/// assert_eq!(p.push(10), Ok(()));
+/// assert_eq!(p.push(20), Err(PushError(20)));
+///
+/// assert!(!p.is_empty());
+/// assert!(p.is_full());
+/// ```
+pub struct Producer<T> {
+    /// The inner representation of the queue.
+    inner: Arc<Inner<T>>,
+
+    /// A copy of `inner.head` for quick access.
+    ///
+    /// This value can be stale and sometimes needs to be resynchronized with `inner.head`.
+    head: Cell<usize>,
+
+    /// A copy of `inner.tail` for quick access.
+    ///
+    /// This value is always in sync with `inner.tail`.
+    tail: Cell<usize>,
+}
+
+unsafe impl<T: Send> Send for Producer<T> {}
+
+impl<T> Producer<T> {
+    /// Attempts to push an element into the queue.
+    ///
+    /// If the queue is full, the element is returned back as an error.
```

Comment:
This may sound silly, but I'd be happy if this behavior-when-full would be stated a bit more prominently in the module docs.

Reply:
Sorry, not sure what you mean exactly. Would you be happy if the first sentence said the following?

```rust
/// Attempts to push an element into the queue, returning it back on failure.
```

Reply:
Sorry if I was unclear, I meant for the whole spsc module (rustdoc and/or README). Something like "a bounded SPSC queue that allocates a fixed-capacity buffer on construction, preventing insertions when full".